// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// The protocol buffers below are used to store information about instructions.

syntax = "proto3";

package exegesis;

import "exegesis/proto/instruction_encoding.proto";
import "exegesis/proto/microarchitecture.proto";
import "exegesis/proto/registers.proto";
import "exegesis/proto/x86/encoding_specification.proto";

// The Intel documentation referred to here can be found at:
// http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf

// Contains information about a single operand of a function.
message InstructionOperand {
  // The name of the operand, as listed by the vendor.
  string name = 4;

  // Human-readable description of the operand, as provided by the vendor,
  // in Markdown format.
  string description = 9;

  // Concrete value to be set for this operand, if applicable.
  bytes value = 10;

  // Defines the possible addressing modes of an operand.
  // clang-format off
  enum AddressingMode {
    // The operand can be using any addressing mode.
    ANY_ADDRESSING_MODE = 0x0;

      // The operand is an immediate value, and its value is stored directly
      // in the encoded instruction.
      NO_ADDRESSING = 0x1;

      // The operand can use either direct or indirect addressing using one or
      // more registers whose indices are explicitly encoded in the instruction.
      ANY_ADDRESSING_WITH_FLEXIBLE_REGISTERS = 0x2;

        // The operand uses direct addressing. The operand is the register whose
        // index is encoded in the instruction.
        DIRECT_ADDRESSING = 0x21;

          // The operand uses block addressing. This is used in the block
          // AVX-512 FMA instructions; the register set is divided into blocks
          // of four registers (0-3, 4-7, ...), and the register used as operand
          // picks the whole block.
          BLOCK_DIRECT_ADDRESSING = 0x212;

        // The operand uses indirect addressing. The operand is the memory
        // location addressed by zero or more registers.
        INDIRECT_ADDRESSING = 0x22;

          // The memory is addressed using a base register.
          INDIRECT_ADDRESSING_WITH_BASE = 0x221;

          // The memory is addressed using only a displacement (the address of
          // the memory location is hard-coded in the instruction).
          INDIRECT_ADDRESSING_WITH_DISPLACEMENT = 0x222;

          // The memory is addressed using a base register and a displacement.
          INDIRECT_ADDRESSING_WITH_BASE_AND_DISPLACEMENT = 0x223;

          // The memory is addressed using a base register and an index
          // register.
          INDIRECT_ADDRESSING_WITH_BASE_AND_INDEX = 0x224;

          // The memory is addressed using an index register and a displacement.
          INDIRECT_ADDRESSING_WITH_INDEX_AND_DISPLACEMENT = 0x225;

          // The memory is addressed using a base register, an index register
          // and a displacement.
          INDIRECT_ADDRESSING_WITH_BASE_DISPLACEMENT_AND_INDEX = 0x226;

          // The memory is addressed using the x86-64 VSIB addressing scheme.
          INDIRECT_ADDRESSING_WITH_VSIB = 0x227;

          // The memory is addressed using instruction pointer register.
          INDIRECT_ADDRESSING_WITH_INSTRUCTION_POINTER = 0x228;

      // The operand uses either direct or indirect addressing using a register
      // that is fixed for the given instruction class, and it can't be changed
      // in the encoding.
      ANY_ADDRESSING_WITH_FIXED_REGISTERS = 0x3;

        // The memory is addressed using the source index register RSI.
        INDIRECT_ADDRESSING_BY_RSI = 0x31;

        // The memory is addressed using the destination index register RDI.
        INDIRECT_ADDRESSING_BY_RDI = 0x32;

      // The operand uses the same rules for computing a memory location as the
      // indirect addressing operands, but it does not access the data at the
      // location. This is used for example by the LEA instructions on x86-64.
      LOAD_EFFECTIVE_ADDRESS = 0x4;
  }
  // clang-format on

  // Defines the possible ways of encoding an operand.
  // clang-format off
  enum Encoding {
    // The operand can be using any encoding. This value should never be used in
    // practice.
    /* ABSTRACT */ ANY_ENCODING = 0;
      // The operand is a register operand of an x86-64 instruction.
      /* ABSTRACT */ X86_REGISTER_ENCODING = 0x1;
        // The operand is encoded in the opcode of the instruction.
        OPCODE_ENCODING = 0x11;

        // The operand is encoded using the ModR/M byte.
        /* ABSTRACT */ MODRM_ENCODING = 0x12;

          // The operand is encoded in the modrm.reg bits.
          MODRM_REG_ENCODING = 0x121;

          // The operand is encoded in the modrm.rm bits, and optionally also in
          // the SIB byte.
          MODRM_RM_ENCODING = 0x122;

        // The operand is an operand of a VEX/EVEX instruction, and it is
        // encoded in one of the ways introduced with VEX encoding.
        /* ABSTRACT */ VEX_ENCODING = 0x13;

          // The operand is encoded in the vex.vvvv bits of the VEX/EVEX prefix.
          VEX_V_ENCODING = 0x131;

          // The operand is encoded in the VEX operand suffix.
          VEX_SUFFIX_ENCODING = 0x132;

        // The operand is a part of an EVEX instruction, and it is encoded in
        // one of the ways introduced with the EVEX encoding.
        /* ABSTRACT */ EVEX_ENCODING = 0x14;

          // The operand is the masking register and it is encoded in the
          // evex.aaa bits of the EVEX prefix.
          EVEX_MASK_OPERAND_ENCODING = 0x141;

      // The operand is an implicit operand of the instruction.
      IMPLICIT_ENCODING = 0x2;

          X86_REGISTER_EAX = 0x21;
          X86_REGISTER_EBX = 0x22;
          X86_REGISTER_RAX = 0x23;
          X86_REGISTER_RBX = 0x24;
          X86_REGISTER_RCX = 0x25;
          X86_REGISTER_RDX = 0x26;

      // The operand is an immediate value. Its value is stored as a part of the
      // instruction encoding.
      IMMEDIATE_VALUE_ENCODING = 0x3;

      // The operand is encoded using the VSIB scheme.
      VSIB_ENCODING = 0x4;

      // The operand is a static property of the instruction that is represented
      // as an operand in the syntax of the instruction.
      X86_STATIC_PROPERTY_ENCODING = 0x5;
  }
  // clang-format on

  // The addressing mode used (allowed) by the operand.
  AddressingMode addressing_mode = 1;

  // The encoding of the operand.
  Encoding encoding = 2;

  // Contains information about the data type of an operand.
  //
  // Note that this is the logical data type the instruction uses to perform the
  // operation, which may be a different size to the location the value is read
  // from or written to, e.g. instructions which implicitly sign- or zero-extend
  // operands or results, opmask registers which are only partially used for
  // smaller vector sizes.
  message DataType {
    enum Kind {
      UNKNOWN = 0;

      FLOATING_POINT = 1;

      // An aggregate value laid out according to some specified format. This is
      // usually only used by a few specialised instructions, e.g. LGDT in x86.
      STRUCT = 4;

      INTEGER = 5;

      // Values which are used purely as opaque collections of bits. For
      // example, the operands to instructions which just move data between
      // registers or memory.
      NONE = 6;
    }

    // The type of the type.
    Kind kind = 1;

    // The number of bits used to represent the value. If the value is a vector,
    // this is only the size of a single element of the vector.
    int32 bit_width = 2;

    // If this is a vector value, the number of elements contained in the
    // vector. Each element is typed according to the fields above. For example,
    // the following DataType represents a 128-bit vector value containing 4
    // signed 32-bit ints:
    //    kind: SIGNED_INTEGER
    //    bit_width: 32
    //    vector_width: 4
    int32 vector_width = 3;
  }

  // The data type of the operand.
  DataType data_type = 7;

  // The size of the value in bits. This may differ from the size in data_type
  // in that it represents the "physical size" rather than the "logical size".
  // "Logical size" refers to the size of the data type that the instruction
  // operates on. "Physical size" refers to the size of the location the operand
  // is read from / written to. These may be different in some cases, such as
  // when a value is read from a register but only the low bits are used.
  int32 value_size_bits = 3;

  // This represents tags to mark masking (e.g. {k1}{z} or {k}), or specific
  // support (e.g. {er} or {sae}).
  message Tag {
    // Name, e.g. "k1", "er", ...
    string name = 1;
  }
  repeated Tag tags = 5;

  // How the operand is used: read, write, or both.
  enum Usage {
    USAGE_UNKNOWN = 0;
    USAGE_READ = 1;
    USAGE_WRITE = 2;
    USAGE_READ_WRITE = 3;
  }
  Usage usage = 6;

  // Holds the register_class to which the operand belongs.
  RegisterProto.RegisterClass register_class = 8;
}

// Contains a list of measurements of the CPU events like the number of uoos
// issued, where they have been dispatched, and how many cycles were spent.
// The values are normalized to one instruction.
message ObservationVector {
  message Observation {
    // The name of the event. Added by PerfSubsystem::PopulateEventLists().
    string event_name = 1;

    // The measured value corresponding to the event.
    double measurement = 2;
  }
  repeated Observation observations = 1;
}

message InstructionFormat {
  // This is the key field for an instruction as there can be several
  // instructions with the same name, but differing opcodes, depending on the
  // addressing. There can also be instructions with different names and the
  // same opcode.
  string mnemonic = 1;

  // The information about the operands of the instructions. The operands are
  // listed in the order in which they appear in the vendor syntax
  // specification.
  repeated InstructionOperand operands = 3;
}

// Contains information about a single instruction.
// Next index to use: 34
message InstructionProto {
  // A human-readable description of what the instruction actually does.
  // See the abovementioned Intel document, sections 3.1 and later.
  string description = 1;

  // This is the key field for an instruction as there can be several
  // instructions with the same name, but differing opcodes, depending on the
  // addressing. There can also be instructions with different names and the
  // same opcode.
  string llvm_mnemonic = 2;

  // The instruction as defined in the manuals of the chip vendor. Note that
  // this field serves as a documentation/comment rather than for practical use,
  // and it might not compile with LLVM. To emit this instruction with LLVM, use
  // the data from 'syntax' instead.
  // There may be more than one instance of vendor syntax in case of
  // instructions that have multiple mnemonics. The most notable example in the
  // x86-64 world are the conditional instructions, where "different" conditions
  // are tested using the same flag. For example, "not equal" and "not zero"
  // boil down to the same condition "zero flag == 0".
  repeated InstructionFormat vendor_syntax = 3;

  // The instruction in the LLVM variant of the assembly syntax used by the
  // chip vendor.
  InstructionFormat syntax = 4;

  // The instruction in the AT&T syntax of LLVM.
  InstructionFormat att_syntax = 5;

  // Contains the name of the instruction set extension the instruction belongs
  // to, for example "SSE2" or "AVX". Note that this is a string and not an enum
  // because we don't want this proto to depend on a particular CPU
  // infrastructure.
  // This can also hold combinations for feature names: "A && B" and "A || B"
  // respectively denote that the instruction requires both or any of features
  // A and B.
  string feature_name = 17;

  // Whether the instruction is available in 64-bit mode. This is not the
  // case for example for AAA.
  // TODO(bdb): use an enum to encode the reason why the instruction is not
  // available in 64-bit mode.
  bool available_in_64_bit = 18;

  // True if the instruction is usable in "legacy" mode.
  // TODO(bdb): use an enum to encode the reason why the instruction is not
  // usable in legacy mode.
  bool legacy_instruction = 19;

  // The encoding scheme for the instruction, as defined in the hereabove
  // mentioned Intel manual, section 2.1.5.
  string encoding_scheme = 20;

  // The minimum protected mode. The exact meaning depends on the
  // architecture (e.g. x86-64 has 4 levels [0-3] and ARM has 8). The default is
  // to assume the least privileged mode.
  int32 protection_mode = 29;

  // The encoding specification of the instruction, as provided by the designer
  // of the CPU instruction set. The exact format of the value is
  // platform-dependent:
  // * For x86 and x86-64, the encoding specification language used in the
  //   Intel(r) 64 and IA-32 Architectures Software Developer's Manual.
  string raw_encoding_specification = 22;

  // Names of the register operands that are read by the instruction even if
  // they are not explicitly listed in the encoded form or the assembly code.
  repeated string implicit_input_operands = 23;

  // Names of the register operands that are written by the instruction even if
  // they are not explicitly listed in the encoded form or the assembly code.
  repeated string implicit_output_operands = 24;

  // The parsed binary encoding specification. Which field is used depends on
  // the platform to which the instruction belongs.
  oneof encoding_specification {
    // x86 specific encoding.
    exegesis.x86.EncodingSpecification x86_encoding_specification = 30;

    // This specification is used for ARM and POWER architectures.
    exegesis.FixedSizeInstructionLayout fixed_size_encoding_specification = 31;
  }

  // The index of the instruction group this instruction belongs to.
  uint32 instruction_group_index = 32;

  // The leaf functions of this instruction. Only meaningful for SGX
  // instructions, specifically, ENCLU, ENCLV, ENCLS.
  repeated InstructionProto leaf_instructions = 33;
}

// Contains information about a group of instructions. (e.g. for Intel, this
// corresponds to a full table and associated description/algorithm
// specification/... in the SDM)
message InstructionGroupProto {
  // The group the instruction belong to. For example, VADDPS belongs to ADDPS.
  string name = 1;

  // A short description of the group.
  string short_description = 4;

  // A human-readable description of what the group of instructions actually
  // does.
  string description = 2;

  // A human-readable description of which flags are affected and how.
  message FlagsAffected {
    string content = 1;
  }

  repeated FlagsAffected flags_affected = 3;
}

// Stores information about the source of an instruction set, for debugging.
message InstructionSetSourceInfo {
  // The source used to create the instruction set.
  string source_name = 1;

  // Key/Value entries for metadata.
  message MetadataEntry {
    string key = 1;
    string value = 2;
  }
  repeated MetadataEntry metadata = 2;
}

message InstructionSetProto {
  // There might be several sources. For example historically the Intel SDM was
  // distributed as two files with instructions A-M in Volume 2A and
  // instructions N-Z in volume 2B. When merging instruction sets, make sure
  // that there are no duplicate instructions.
  repeated InstructionSetSourceInfo source_infos = 3;

  repeated InstructionProto instructions = 4;

  repeated InstructionGroupProto instruction_groups = 5;
}

// The MicroOperation message describes a single micro-operation of the
// instruction. It describes the micro-operation in terms of the set of
// execution ports on which the micro-operation can be executed,
// the latency and the dependencies from the other micro-operations of
// the instruction. When executing a given micro-operation, the CPU picks
// one of the execution ports and sends the micro-operation there.
message MicroOperationProto {
  PortMaskProto port_mask = 1;
  int32 latency = 2;
  repeated int32 dependencies = 3;
  bool double_pumped = 4;
  string likely_execution_unit = 5;
}

// The itinerary for an instruction.
message ItineraryProto {
  // This is useful to be able to read stored files more easily. For debug only.
  string llvm_mnemonic = 1;

  // The latency is the number of cycles, after which the output of the
  // instruction is ready, and any other instruction that uses this
  // output can be sent for execution.

  // The minimum latency for the instruction (for example for simple cases).
  int32 min_latency = 6;

  // The maximum latency for the instruction. For most of the basic instructions
  // the minimum latency is equal to the maximum latency.
  int32 max_latency = 7;

  // For complex instructions like ENTER or REP CMPS, the latency can be
  // proportional to the number of bytes treated by the instruction.
  double proportional_latency_per_byte = 8;

  // True if the latency is approximate. This is the case for complex
  // micro-coded instructions.
  bool latency_is_approximate = 9;

  // The minimum throughput for the instruction (for example for simple cases).
  int32 min_throughput = 10;

  // The maximum throughput for the instruction. For most of the basic
  // instructions the minimum throughput is equal to the maximum throughput.
  int32 max_throughput = 11;

  // For complex instructions like ENTER or REP CMPS, the throughput can be
  // proportional to the number of bytes treated by the instruction.
  double proportional_throughput_per_byte = 12;

  // The number of micro-ops in the unfused domain, i.e. before macro-fusion
  // occurs.
  int32 num_uops_unfused_domain = 13;

  // The number of micro-ops in the fused domain, i.e. after macro-fusion
  // occurs. For most instructions, num_uops_unfused_domain >=
  // num_uops_fused_domain.
  int32 num_uops_fused_domain = 14;

  // The following Boolean is true when the instruction is issued in
  // a standard way to the different execution units, by opposition to
  // micro-coded instructions.
  bool standard_execution = 15;

  // An itinerary is an array of MicroOperations describing the
  // different steps that must be performed to execute the instruction.
  repeated MicroOperationProto micro_ops = 16;

  // Observation vector for throughput measurements.
  ObservationVector throughput_observation = 27;

  // Observation vector for latency measurements.
  ObservationVector latency_observation = 28;
}

// A set of instruction itineraries. This is always specific to a given
// microarchitecture.
message InstructionSetItinerariesProto {
  string microarchitecture_id = 1;
  repeated ItineraryProto itineraries = 2;
}

// Represents a set of per-model itineraries for an instruction set. The
// microarchitecture-specific information is kept separate from the ISA which is
// common to all microarchitectures.
message ArchitectureProto {
  // e.g. "x86_64".
  string name = 5;

  // The unannotated instruction set after applying transforms.
  InstructionSetProto instruction_set = 1;

  // Itineraries, per micro-architecture.
  repeated InstructionSetItinerariesProto per_microarchitecture_itineraries = 3;

  // The list of registers available in the instruction set.
  RegisterSetProto register_set = 6;
}

// Allows defining translation of instruction mnemonics from one or more source
// mnemonics to a target mnemonic.
//
// This is used to compensate for LLVM opcodes that do not correspond to a real
// x86 instruction and that are fixed by instruction lowering long after the
// instruciton scheduling pass. An example of such translation is the
// translation of MOV32ri64 (a LLVM-specific instruction) to MOV32ri (a valid
// x86 and x86-64 instruction).
message InstructionMnemonicTranslationProto {
  repeated string source_mnemonics = 1;
  string target_mnemonic = 2;
}

message InstructionMnemonicTranslationListProto {
  repeated InstructionMnemonicTranslationProto translations = 1;
}
